src/math/NewtonRaphson.h

```diff
@@ -0,0 +1,84 @@
+#pragma once
+
+#include "MathException.h"
+#include "math/LineSearch.h"
+
+namespace NuTo
+{
+namespace NewtonRaphson
+{
+
+//! @brief custom exception for the newton algorithm
+class NoConvergence : public NuTo::MathException
+{
+public:
+    NoConvergence(const std::string& caller, const std::string& message)
+        : MathException(caller, message)
+    {
+    }
+};
+
+//! @brief "Solver" for scalar values
+struct DoubleSolver
+{
+    static double Solve(double dr, double r)
+    {
+        return r / dr;
+    }
+};
+
+//! @brief problem definition
+//! @tparam TR residual function
+//! @tparam TDR derivative of the residual function
+//! @tparam TNorm function that calculates the norm of the residual TTol = TNorm(TR)
+//! @tparam TTol tolerance
+//! @tparam TInfo Info function that takes (int, return type of TR, return type of TR)
+template <typename TR, typename TDR, typename TNorm, typename TTol, typename TInfo>
+struct Problem
+{
+    TR ResidualFunction;
+    TDR DerivativeFunction;
+    TNorm NormFunction;
+    TTol mTolerance;
+    TInfo InfoFunction;
+};
+
+//! @brief defines the problem, basically just to enable automatic template deduction. If you
+//! create a Problem directly, you'll have to specify each template parameter. This methods avoids it.
+template <typename TR, typename TDR, typename TNorm, typename TTol, typename TInfo = VoidInfo>
+auto DefineProblem(TR residual, TDR derivative, TNorm norm, TTol tolerance, TInfo info = VoidInfo())
```

Comment:
This _work around_ will be gone in c++17 with [class template deduction](http://en.cppreference.com/w/cpp/language/class_template_deduction).